src/lib/support/BytesCircularBuffer.cpp

```diff
@@ -0,0 +1,158 @@
+    }
+    else
+    {
+        size_t firstPiece  = mCapacity - start;
+        size_t secondPiece = length - firstPiece;
+        ::memcpy(dest, mStorage + start, firstPiece);
+        ::memcpy(dest + firstPiece, mStorage, secondPiece);
+    }
+}
+
+void BytesCircularBuffer::Write(const uint8_t * source, size_t length)
+{
+    // Always reserve 1 byte to prevent mDataStart == mDataEnd because then it would be
+    // ambiguous whether we have 0 bytes or mCapacity bytes stored.
+    VerifyOrDie(mCapacity - StorageUsed() - 1 >= length);
+
+    if (mCapacity - mDataEnd >= length)
+    {
+        ::memcpy(mStorage + mDataEnd, source, length);
+        mDataEnd = Advance(mDataEnd, length); // Use advance in case that mDataEnd wrap to 0
+    }
+    else
+    {
+        size_t firstPiece  = mCapacity - mDataEnd;
+        size_t secondPiece = length - firstPiece;
+        ::memcpy(mStorage + mDataEnd, source, firstPiece);
+        ::memcpy(mStorage, source + firstPiece, secondPiece);
+        mDataEnd = secondPiece;
+    }
+}
+
+void BytesCircularBuffer::Drop(size_t length)
+{
+    VerifyOrDie(StorageUsed() >= length);
+    mDataStart = Advance(mDataStart, length);
+}
+
+size_t BytesCircularBuffer::StorageUsed() const
+{
+    if (mDataStart <= mDataEnd)
+    {
+        return mDataEnd - mDataStart;
+    }
+    else
+    {
+        return mCapacity + mDataEnd - mDataStart;
+    }
+}
+
+CHIP_ERROR BytesCircularBuffer::Push(const ByteSpan & payload)
+{
+    size_t maxStorageAvailable = mCapacity - (sizeof(SizeType) + 1);
+    size_t length              = payload.size();
+    if (length > maxStorageAvailable)
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    if (length > std::numeric_limits<SizeType>::max())
+        return CHIP_ERROR_INVALID_ARGUMENT;
+
+    // Free up space until there is enough space.
+    while (length > maxStorageAvailable - StorageUsed())
+    {
+        VerifyOrDie(Pop() == CHIP_NO_ERROR);
+    }
+
+    SizeType size = static_cast<SizeType>(length);
+    Write(reinterpret_cast<uint8_t *>(&size), sizeof(size));
+    Write(payload.data(), length);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BytesCircularBuffer::Pop()
+{
+    if (IsEmpty())
+        return CHIP_ERROR_INCORRECT_STATE;
+
+    size_t length = GetFrontSize();
+    Drop(sizeof(SizeType));
+    Drop(length);
+
+    return CHIP_NO_ERROR;
+}
+
+bool BytesCircularBuffer::IsEmpty() const
+{
+    return StorageUsed() == 0;
+}
+
+size_t BytesCircularBuffer::GetFrontSize() const
+{
+    if (IsEmpty())
+        return 0;
+
+    SizeType length;
+    Read(reinterpret_cast<uint8_t *>(&length), sizeof(length), 0 /* offset */);
+    return length;
+}
+
+CHIP_ERROR BytesCircularBuffer::ReadFront(MutableByteSpan & dest) const
+{
+    if (IsEmpty())
+        return CHIP_ERROR_INCORRECT_STATE;
+
+    size_t length = GetFrontSize();
+    if (dest.size() < length)
+        return CHIP_ERROR_INVALID_ARGUMENT;
+
+    dest = dest.SubSpan(0, length);
```

Comment:
`dest.reduce_size(length)` might be more idiomatic... hard to tell.  OK by me either way.